internal/output/result.go

```diff
@@ -0,0 +1,126 @@
+package output
+
+import (
+	"encoding/json"
+
+	"github.com/google/osv-scanner/pkg/models"
+	"golang.org/x/exp/slices"
+)
+
+type pkgWithSource struct {
+	Package models.PackageInfo
+	Source  models.SourceInfo
+}
+
+// Custom implementation of this unique set map to allow it to serialize to JSON
+type pkgSourceSet map[pkgWithSource]struct{}
+
+func (pss *pkgSourceSet) MarshalJSON() ([]byte, error) {
+	res := []pkgWithSource{}
+
+	for v := range *pss {
+		res = append(res, v)
+	}
+
+	return json.Marshal(res)
+}
+
+func (pss *pkgSourceSet) UnmarshalJSON(data []byte) error {
+	aux := []pkgWithSource{}
+	if err := json.Unmarshal(data, &aux); err != nil {
+		return err
+	}
+
+	*pss = make(pkgSourceSet)
+	for _, pws := range aux {
+		(*pss)[pws] = struct{}{}
+	}
+
+	return nil
+}
+
+type groupedVuln struct {
```

Comment:
nit: let's name this `sarifFinding` or something similar since it's very specific to that? Probably have some of the terminology a bit wrong here, but hopefully you get the idea :) 

`groupedVuln` on its own is otherwise a bit confusing? Unless you think this is something that generalizes to other output formats?